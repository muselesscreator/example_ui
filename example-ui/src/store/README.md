# Redux Store
## What's in this module?

### index.js (default store import!)
This is the default import point for the redux store in the majority of the project, imported as 
```js
import {actions, selectors, thunkActions, global_selectors} from 'store';
```

### store.js
This file is the source for the above index file, specifically separated out so that there can be a family of global selectors and thunkActions without introducing circular dependencies.
 
### global_selectors
This is the home for selectors that want to access multiple store modules.

### thunkActions
This is the home for thunkActions that want to access multiple store modules, particularly between the argus websocket api and the gui store module.
