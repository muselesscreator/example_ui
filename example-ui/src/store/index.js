/* Main import point for redux store.
 * The base versions of pretty much everything exported from here is actually 
 * defined in './store.js'.  This import trick is so that we can have a set of
 * global selectors with access to the finalized selectors without introducing
 * import loops.
 *
 * The only ting actually changed here is the thunkActions, which merge in a
 * local collection to those imported via connected modules.
 */

import {
  actions,
  reducer,
  selectors,
  thunkActions,
} from './store';

/* This chunk of code is to better enable debugging of the UI in development mode.
 * logs out all actions, thunkActions, selectors, and global_selectors and provides
 * window-level access to them.
 *
 * This is split down a bit as past a certain point, js logging stops noting nested 
 * modules as log-able dicts.
 */
// TODO: Lock this to admin mode?
if (window.name !== 'nodejs') {
  console.log({
    actions,
    selectors,
  });
}

export {
  actions,
  reducer as default,
  selectors,
  thunkActions,
};
