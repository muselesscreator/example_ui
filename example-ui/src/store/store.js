import { connectReducers } from 'js-libs/build/redux';

import * as simple_store from './modules/simple_store';

const {
  thunkActions: _thunkActions, actions, reducer, selectors
} = connectReducers({
  simple_store,
});

export {
  actions,
  reducer,
  selectors
};
export const thunkActions = { ..._thunkActions };
