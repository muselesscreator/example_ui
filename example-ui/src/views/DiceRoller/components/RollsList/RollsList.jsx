/** @module */
import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { actions, selectors } from 'store';
import { Button } from 'components';

import './RollsList.scss';

/**
 * Simple display component for the D&D dice rolls list.
 * Provides a formatted list of roll strings, with a button at the bottom
 * allowing the user to clear all registered rolls.
 */
export const RollsList = ({ clearRolls, rolls, maxRoll, numRolls }) =>
  <div className="rolls-list">
    <div className="rolls-container">
      { rolls.map((roll, index) => (
        <div className="roll-entry" key={index}>
          {roll}
        </div>
      )) }
      Number of Rolls: {numRolls}
    </div>
    <div className="action-bar">
      <Button
        className="clear-rolls-btn"
        onClick={clearRolls}
      >
        Clear
      </Button>
    </div>
    
  </div>

RollsList.propTypes = {
  // redux
  rolls: PropTypes.arrayOf(PropTypes.string),
  clearRolls: PropTypes.func,
};

export const mapStateToProps = (state) => ({
  rolls: selectors.simple_store.rolls(state),
  numRolls: selectors.simple_store.numRolls(state),
});

export const mapDispatchToProps = {
  clearRolls: actions.simple_store.clearRolls,
};

export default connect(mapStateToProps, mapDispatchToProps)(RollsList);
