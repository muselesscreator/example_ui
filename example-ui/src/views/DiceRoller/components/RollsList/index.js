import RollsList from './RollsList';

export default RollsList;
