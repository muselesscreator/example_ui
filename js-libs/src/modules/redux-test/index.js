import testActions from './actions';
import * as components from './components';
import testReducer, { tests as reducerTests } from './reducers';
import testSelectors from './selectors';

export {
  testActions,
  components,
  testReducer,
  reducerTests,
  testSelectors,
}
