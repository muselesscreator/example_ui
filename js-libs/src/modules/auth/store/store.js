import { action, mkReducer, simpleSelectors } from 'modules/redux';
import { StrictDict } from 'modules/common';

import Auth from '../comms';
import * as thunkActions from './thunkActions';

export const initialState = {
  user: null,
  parseHash: null,
};

export const actions = StrictDict({
  userProfileLoaded: action('USER_PROFILE_LOADED', ['user']),
  handleAuthCallback: action('HANDLE_AUTH_CALLBACK', ['parseHash']),
});

export const reducer = mkReducer(initialState, {
  [actions.userProfileLoaded.type]: (state, { user }) => ({ ...state, user }),
  [actions.handleAuthCallback.type]: (state, { parseHash }) => ({ ...state, parseHash }),
});

export const selectors = {
  ...simpleSelectors(initialState),
  isAuthenticated: ({ user }) => user !== null,
  userName: ({ user }) => user !== null ? Auth.selectors.userName(user) : null,
};

export { thunkActions };
