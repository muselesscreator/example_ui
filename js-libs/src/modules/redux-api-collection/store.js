/** @module */
import { action, mkReducer } from 'modules/redux';
import { StrictDict } from 'modules/common';
import { connectedSelectors } from './connectedSelectors';

import * as module from './store';

export const simpleStore = ({
  key,
  actions,
  commsSelectors,
  reducerFunctions,
  selectors,
}) => ({
  initialState: module.initialState(key),
  actions: module.actions(key, actions),
  reducer: module.reducer(key, reducerFunctions),
  selectors: module.mkSelectors(key, { ...commsSelectors, ...selectors }),
})


export const initialState = (key) => ({ [key]: {} });

/**
 * load action creator
 * @param {string} key - module key
 * @return {object} - load action creator object
 */
export const loadAction = (key) => action('LOAD', [key]);

export const actions = (key, extraActions) => StrictDict({
  load: module.loadAction(key),
  ...extraActions,
});

export const handleLoad = (key) => (
  (state, action) => ({
    ...state,
    [key]: action[key],
  })
);

/**
 * @callback reducer
 * @param {object} state - redux state
 * @param {object} action - redux action being handled
 * @return {object} - new redux state
 */

/**
 * generates a reducer given a module key and an object of additional/overloaded
 * actions handler functions.
 * @param {string} key - module key
 * @param {object} [reducerFunctions] - extra action handlers and/or an overload for
 *   handling the load action
 * @return {reducer} - comms-connected reducer function
 */
export const reducer = (key, reducerFunctions) => mkReducer(
  module.initialState(key),
  { 
    [module.loadAction(key).type]: module.handleLoad(key),
    ...reducerFunctions,
  }
);

/**
 * @function
 * Generates selectors.  In a function so that conenctedSampleSelectors
 * can be mocked out for testing.
 */
export const mkSelectors = (key, commsSelectors) => ({
  list: (state) => state[key],
  /**
   * Return a given order based on its index/order_id
   * @param {string} index - order index
   */
  atIndex: (state, index) => state[key][index],
  ...connectedSelectors({ ...commsSelectors }, key),
});

export default simpleStore;
