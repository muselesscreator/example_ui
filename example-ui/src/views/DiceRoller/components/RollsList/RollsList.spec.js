import { components } from 'js-libs/build/redux-test';

import * as module from './RollsList';

import { selectors, actions } from 'store';

jest.mock('store', () => ({
  selectors: {},
  actions: {
    simple_store: {
      clearRolls: jest.fn().mockName('clearRolls'),
    },
  },
}));

describe('RollsList component', () => {
  describe('Component', () => {
    const props = {
      clearRolls: jest.fn().mockName('clearRolls'),
      rolls: ['some', 'rolls'],
    };
    components.simpleTest('smoke test', module.RollsList, props);
  });
  describe('mapSateToProps', () => {
    components.testStateToProps(
      module.mapStateToProps,
      selectors,
      {
        rolls: 'simple_store.rolls',
      }
    );
  });
  describe('mapDispatchToProps', () => {
    const { mapDispatchToProps: props } = module;
    test('clearRolls - linked to simple_store action', () => {
      expect(props.clearRolls).toEqual(actions.simple_store.clearRolls);
    });
  });
});

