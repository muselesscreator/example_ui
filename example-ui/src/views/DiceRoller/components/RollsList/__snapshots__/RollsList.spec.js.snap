// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RollsList component Component :smoke test 1`] = `
<div
  className="rolls-list"
>
  <div
    className="rolls-container"
  >
    <div
      className="roll-entry"
      key="0"
    >
      some
    </div>
    <div
      className="roll-entry"
      key="1"
    >
      rolls
    </div>
  </div>
  <div
    className="action-bar"
  >
    <Button
      className="clear-rolls-btn"
      onClick={[MockFunction clearRolls]}
    >
      Clear
    </Button>
  </div>
</div>
`;
